### PR TITLE
Add generated protocol types

### PR DESCRIPTION
**Pull request by zardoy**
Add intellisense for packets:

Unfortunately in TS no way to display optionality or jsdoc for overloaded methods, its only possible with js proxied objects.

Alternatively, data can be generated in minecraft-data and reused here as type.

If everything is ok, I'll add the same for events (receive from server data)

### CODE REVIEW

File: scripts/genTypes.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
+const fs = require('fs')
+const path = require('path')
+const mcData = require('minecraft-data')
+// const dataPaths = require('../minecraft-data/data/dataPaths.json')
```
Comment (rom1504) [inline scripts/genTypes.js:4]: commented code should be removed

File: scripts/genTypes.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+const fs = require('fs')
```
Comment (rom1504) [inline scripts/genTypes.js:1]: this logic should be tested, there are a lot of edge cases so it's probably not fully working as is, tests will make it possible to fix it

node protodef has test values

File: scripts/genTypes.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  // if (version !== '1.20') continue
+  // const protocolPath = path.join(__dirname, `../minecraft-data/data/pc/${version}/protocol.json`);
+  // if (!fs.existsSync(protocolPath)) continue
+  // console.log('version', version)
+  // const protocol = require(protocolPath)
+  const protocol = mcData(version).protocol
+
+  const packetKeys = []
+  // from client
+  const packets = Object.entries(protocol.play.toServer.types).map(([name, [type, props]]) => {
```
[the remark this answers is not in the document]
Comment (zardoy) [reply scripts/genTypes.js:25]: but [noprotodefde-protodef ](https://github.com/ProtoDef-io/node-protodef/blob/master/package.json) doesn't have the minecraft-data dependency, what's your proposing workflow?

[the remark this answers is not in the document]
Comment (rom1504) [reply scripts/genTypes.js:25]: I think the code generating typescript types for protodef types should be generic to any protodef type.
Node protodef would then expose a function for generating such typescript types and then node-minecraft-protocol would use it

**Comment by rom1504** [comment]
thanks for the pr; but as said in comments this should live in node-protodef and be more generic

node-minecraft-protocol relies on node-protodef for the definition and implementation of its serialization functions

**Comment by zardoy** [comment]
I'm really interested in this. I do hope to not forget to do this, The last time I looked at it I realized I need more time to implement it there... for now I will just use generated files: https://github.com/zardoy/prismarine-web-client/blob/d36bec02bfb7222ef8ea1b0503371bf16812a9e2/src/generatedServerPackets.ts its really handy
